source/mqtt_client_connection.c

```diff
@@ -692,14 +698,38 @@ PyObject *aws_py_mqtt_client_connection_subscribe(PyObject *self, PyObject *args
         suback_callback);
 
     if (msg_id == 0) {
-        Py_DECREF(callback);
-        Py_DECREF(suback_callback);
+        Py_XDECREF(callback);
+        Py_XDECREF(suback_callback);
         return PyErr_AwsLastError();
     }
 
     return PyLong_FromUnsignedLong(msg_id);
 }
 
+PyObject *aws_py_mqtt_client_connection_on_message(PyObject *self, PyObject *args) {
+    (void)self;
+
+    PyObject *impl_capsule;
+    PyObject *callback;
+    if (!PyArg_ParseTuple(args, "OO", &impl_capsule, &callback)) {
+        return NULL;
+    }
+
+    struct mqtt_connection_binding *py_connection =
+        PyCapsule_GetPointer(impl_capsule, s_capsule_name_mqtt_client_connection);
+    if (!py_connection) {
+        return NULL;
+    }
+
+    callback = PyWeakref_NewProxy(callback, NULL);
```

Comment:
2 things:
1) not checking for NULL return
2) if this succeeds, we never DECREF the callback, so the proxy will leak. You probably want to store a member on the binding object `PyObject *on_message;` so we can DECREF it in the binding destructor.
    - Also, in this fn, be sure to protect against someone setting on_message multiple times. DECREF the existing value before setting the new callback and INCREFing it